File: bench/kernels/bench_tailcall.js
// bench_tailcall.js - self-recursive tail call, depth 2,000,000 (qjs)
function tailsum(n, acc) {
    if (n === 0) return acc;
    return tailsum(n - 1, acc + n);
}

const n = 2000000;
const t0 = Date.now();
const result = tailsum(n, 0);
const elapsed = Date.now() - t0;
print("result: " + result);
print("elapsed: " + elapsed + " ms");

File: bench/kernels/bench_tailcall.rs
// bench_tailcall.rs - self-recursive tail call, depth 2,000,000
use std::time::Instant;

fn tailsum(n: i64, acc: i64) -> i64 {
    if n == 0 {
        return acc;
    }
    tailsum(n - 1, acc + n)
}

fn main() {
    let n: i64 = 2_000_000;
    let t0 = Instant::now();
    let result = tailsum(n, 0);
    let elapsed = t0.elapsed().as_millis();
    println!("result: {}", result);
    println!("elapsed: {} ms", elapsed);
}
